File: services/rateLimitMocking.ts
/**
 * Facade for rate limit header processing
 * This isolates mock logic from production code
 */

import {
  applyMockHeaders,
  checkMockFastModeRateLimit,
  getMockHeaderless429Message,
  getMockHeaders,
  isMockFastModeRateLimitScenario,
  shouldProcessMockLimits,
} from './mockRateLimits.js'
import { MossenAPIError } from './api/mossenSdk.js'
import { getCanonicalName } from '../utils/model/model.js'

function isFrontierModel(model: string): boolean {
  const normalized = model.toLowerCase()
  const canonical = getCanonicalName(normalized)
  return (
    canonical.includes('mossen-opus') ||
    normalized === 'opus' ||
    normalized.startsWith('opus[')
  )
}

/**
 * Process headers, applying mocks if /mock-limits command is active
 */
export function processRateLimitHeaders(
  headers: globalThis.Headers,
): globalThis.Headers {
  // Only apply mocks for internal testers using the /mock-limits command.
  if (shouldProcessMockLimits()) {
    return applyMockHeaders(headers)
  }
  return headers
}

/**
 * Check if we should process rate limits (either real subscriber or /mock-limits command)
 */
export function shouldProcessRateLimits(isSubscriber: boolean): boolean {
  return isSubscriber || shouldProcessMockLimits()
}

/**
 * Check if mock rate limits should throw a 429 error
 * Returns the error to throw, or null if no error should be thrown
 * @param currentModel The model being used for the current request
 * @param isFastModeActive Whether fast mode is currently active (for fast-mode-only mocks)
 */
export function checkMockRateLimitError(
  currentModel: string,
  isFastModeActive?: boolean,
): MossenAPIError | null {
  if (!shouldProcessMockLimits()) {
    return null
  }

  const headerlessMessage = getMockHeaderless429Message()
  if (headerlessMessage) {
    return new MossenAPIError(
      429,
      { error: { type: 'rate_limit_error', message: headerlessMessage } },
      headerlessMessage,
      // eslint-disable-next-line eslint-plugin-n/no-unsupported-features/node-builtins
      new globalThis.Headers(),
    )
  }

  const mockHeaders = getMockHeaders()
  if (!mockHeaders) {
    return null
  }

  // Check if we should throw a 429 error
  // Only throw if:
  // 1. Status is rejected AND
  // 2. Either no overage headers OR overage is also rejected
  // 3. For Frontier-tier limits, only throw if actually using a Frontier model
  const status = mockHeaders['mossen-ratelimit-unified-status']
  const overageStatus =
    mockHeaders['mossen-ratelimit-unified-overage-status']
  const rateLimitType =
    mockHeaders['mossen-ratelimit-unified-representative-claim']

  // Check if this is a Frontier-tier rate limit.
  const isFrontierLimit = rateLimitType === 'seven_day_opus'

  // Check if current model is a Frontier model (handles variants and aliases).
  const isUsingFrontier = isFrontierModel(currentModel)

  // For Frontier limits, only throw 429 if actually using Frontier.
  // This simulates the real API behavior where fallback to Balanced succeeds.
  if (isFrontierLimit && !isUsingFrontier) {
    return null
  }

  // Check for mock fast mode rate limits (handles expiry, countdown, etc.)
  if (isMockFastModeRateLimitScenario()) {
    const fastModeHeaders = checkMockFastModeRateLimit(isFastModeActive)
    if (fastModeHeaders === null) {
      return null
    }
    // Create a mock 429 error with the fast mode headers
    const error = new MossenAPIError(
      429,
      { error: { type: 'rate_limit_error', message: 'Rate limit exceeded' } },
      'Rate limit exceeded',
      // eslint-disable-next-line eslint-plugin-n/no-unsupported-features/node-builtins
      new globalThis.Headers(
        Object.entries(fastModeHeaders).filter(([_, v]) => v !== undefined) as [
          string,
          string,
        ][],
      ),
    )
    return error
  }

  const shouldThrow429 =
    status === 'rejected' && (!overageStatus || overageStatus === 'rejected')

  if (shouldThrow429) {
    // Create a mock 429 error with the appropriate headers
    const error = new MossenAPIError(
      429,
      { error: { type: 'rate_limit_error', message: 'Rate limit exceeded' } },
      'Rate limit exceeded',
      // eslint-disable-next-line eslint-plugin-n/no-unsupported-features/node-builtins
      new globalThis.Headers(
        Object.entries(mockHeaders).filter(([_, v]) => v !== undefined) as [
          string,
          string,
        ][],
      ),
    )
    return error
  }

  return null
}

/**
 * Check if this is a mock 429 error that shouldn't be retried
 */
export function isMockRateLimitError(error: MossenAPIError): boolean {
  return shouldProcessMockLimits() && error.status === 429
}

/**
 * Check if /mock-limits command is currently active (for UI purposes)
 */
export { shouldProcessMockLimits }

File: crates/mossen-agent/src/services/root/rate_limit_mocking.rs
/// Facade for rate limit header processing.
/// This isolates mock logic from production code.

use super::mock_rate_limits::{
    apply_mock_headers, check_mock_fast_mode_rate_limit, get_mock_headerless_429_message,
    get_mock_headers, is_mock_fast_mode_rate_limit_scenario, should_process_mock_limits,
};

/// Trait for model name resolution
pub trait RateLimitMockingContext: Send + Sync {
    fn get_canonical_name(&self, model: &str) -> String;
}

fn is_frontier_model(ctx: &dyn RateLimitMockingContext, model: &str) -> bool {
    let normalized = model.to_lowercase();
    let canonical = ctx.get_canonical_name(&normalized);
    canonical.contains("mossen-opus")
        || normalized == "opus"
        || normalized.starts_with("opus[")
}

/// Process headers, applying mocks if /mock-limits command is active
pub fn process_rate_limit_headers(
    headers: std::collections::HashMap<String, String>,
) -> std::collections::HashMap<String, String> {
    if should_process_mock_limits() {
        return apply_mock_headers(headers);
    }
    headers
}

/// Check if we should process rate limits (either real subscriber or /mock-limits)
pub fn should_process_rate_limits(is_subscriber: bool) -> bool {
    is_subscriber || should_process_mock_limits()
}

/// Check if mock rate limits should produce a 429 error.
/// Returns Some(error_message) if a 429 should be thrown, None otherwise.
pub fn check_mock_rate_limit_error(
    ctx: &dyn RateLimitMockingContext,
    current_model: &str,
    is_fast_mode_active: Option<bool>,
) -> Option<MockRateLimitError> {
    if !should_process_mock_limits() {
        return None;
    }

    if let Some(headerless_msg) = get_mock_headerless_429_message() {
        return Some(MockRateLimitError {
            status: 429,
            message: headerless_msg,
            headers: std::collections::HashMap::new(),
        });
    }

    let mock_headers = get_mock_headers()?;

    let status = mock_headers.get("mossen-ratelimit-unified-status").map(|s| s.as_str());
    let overage_status = mock_headers
        .get("mossen-ratelimit-unified-overage-status")
        .map(|s| s.as_str());
    let rate_limit_type = mock_headers
        .get("mossen-ratelimit-unified-representative-claim")
        .map(|s| s.as_str());

    let is_frontier_limit = rate_limit_type == Some("seven_day_opus");
    let is_using_frontier = is_frontier_model(ctx, current_model);

    if is_frontier_limit && !is_using_frontier {
        return None;
    }

    // Check for mock fast mode rate limits
    if is_mock_fast_mode_rate_limit_scenario() {
        if let Some(fast_headers) = check_mock_fast_mode_rate_limit(is_fast_mode_active) {
            return Some(MockRateLimitError {
                status: 429,
                message: "Rate limit exceeded".to_string(),
                headers: fast_headers,
            });
        }
        return None;
    }

    let should_throw_429 =
        status == Some("rejected") && (overage_status.is_none() || overage_status == Some("rejected"));

    if should_throw_429 {
        return Some(MockRateLimitError {
            status: 429,
            message: "Rate limit exceeded".to_string(),
            headers: mock_headers,
        });
    }

    None
}

#[derive(Debug, Clone)]
pub struct MockRateLimitError {
    pub status: u16,
    pub message: String,
    pub headers: std::collections::HashMap<String, String>,
}

/// Check if this is a mock 429 error that shouldn't be retried
pub fn is_mock_rate_limit_error(status: u16) -> bool {
    should_process_mock_limits() && status == 429
}
